src/sage/graphs/digraph.py

```diff
@@ -2875,30 +2922,187 @@ def _all_cycles_iterator_vertex(self, vertex, starting_vertices=None, simple=Fal
+          and ``by_weight`` is ``True``, we use the edge label ``l`` as a
+          weight.
+
+        - ``by_weight`` -- boolean (default: ``False``); if ``True``, the edges
+          in the graph are weighted, otherwise all edges have weight 1
+
+        - ``check_weight`` -- boolean (default: ``True``); whether to check that
+          the ``weight_function`` outputs a number for each edge
+
+        - ``report_weight`` -- boolean (default: ``False``); if ``False``, just
+          a cycle is returned. Otherwise a tuple of cycle length and cycle is
+          returned.
+
+        OUTPUT: iterator
+
+        ALGORITHM:
+
+        Given an edge `uv`, this algorithm extracts k-shortest `vu`-path in `G-uv`
+        in increasing length order by using k-shortest path algorithm. Thus, it
+        extracts only simple cycles. See
+        :math:`~sage.graphs.path_enumeration.shortest_simple_paths` for more
+        information.
+
+        EXAMPLES:
+
+            sage: g = graphs.Grid2dGraph(2, 5).to_directed()
+            sage: it = g._all_simple_cycles_iterator_edge(((0, 0), (0, 1), None), report_weight=True)
+            sage: for i in range(5): print(next(it))
+            (2, [(0, 0), (0, 1), (0, 0)])
+            (4, [(0, 0), (0, 1), (1, 1), (1, 0), (0, 0)])
+            (6, [(0, 0), (0, 1), (0, 2), (1, 2), (1, 1), (1, 0), (0, 0)])
+            (8, [(0, 0), (0, 1), (0, 2), (0, 3), (1, 3), (1, 2), (1, 1), (1, 0), (0, 0)])
+            (10, [(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (1, 4), (1, 3), (1, 2), (1, 1), (1, 0), (0, 0)])
+
+        Each edge must have a positive weight::
+
+            sage: g = DiGraph()
+            sage: g.add_edges([('a', 'b', -2), ('b', 'a', 1)])
+            sage: next(g._all_simple_cycles_iterator_edge(('a', 'b', -2), max_length=None, by_weight=True))
+            Traceback (most recent call last):
+            ...
+            ValueError: negative weight is not allowed
+        """
+        if starting_vertices is None:
+            starting_vertices = [edge[0]]
+        if edge[0] not in starting_vertices:
+            raise ValueError("the tail of edge must be one of the starting vertices")
+
+        # First we remove vertices and edges that are not part of any cycle
+        if remove_acyclic_edges:
+            sccs = self.strongly_connected_components()
+            d = {}
+            for id, component in enumerate(sccs):
+                for v in component:
+                    d[v] = id
+            h = copy(self)
+            h.delete_edges((u, v) for u, v in h.edge_iterator(labels=False) if d[u] != d[v])
+        else:
+            h = copy(self)
+        # delete edge
+        h.delete_edge(edge)
+        # delete vertices unnecessary for enumearting cycles
+        if not rooted:
+            for v in starting_vertices:
+                if v < edge[0]:
+                    h.delete_vertex(v)
+                    if v == edge[1]:
+                        return
```

Comment:
I understand your algorithm intends not to output the same cycle more than once. My implemented algorithm also try to do that by deleting vertices (if `rooted=False`). If `rooted=True`, it outputs the same cycle with a different start vertex but I think it is a correct behavior.